Improve table index properties: add new method for selecting index and support multi-column indices

taldcroft opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

NOTE: this is currently draft and I am looking for feedback on the basic idea of the new syntax before doing the docs, what's new, more testing, etc.

#### New method for selecting an index
Currently if multiple indices are defined on a table then the way to choose the index is by supplying the index name as the first element of the item tuple. This looks like below:
```
t = simple_table()
t.add_index("a")
t.add_index(["a", "b"])
t.loc[("a", "b"), (1, 1.0)]
```
This syntax is awkward and not really obvious. It is also potentially ambiguous:
```
>>> t = Table()
>>> t["a"] = ["a", "b", "c"]
>>> t["b"] = [2, 1, 1]
>>> t.add_index(["a", "b"])
>>> t.add_index(["b"])
>>> t.loc["b", 1]  
# Could be key=("b", 1) for primary index on "a", "b" or
# key=1 for index on column "b". The latter is what actually happens.
```
This PR introduces a new method for specifying the index name for the `loc`, `iloc`, and
`loc_indices` index properties.
```
>>> t.loc["b", 1]  # Current syntax now deprecated
WARNING: AstropyDeprecationWarning: Calling `Table.loc/iloc/loc_indices[index_name, item]` to select `item` from index
`index_name` is deprecated. Instead select the index using the method
`Table.loc/iloc/loc_indices.with_index(index_name)[item]`.
 [astropy.table.index]

>>> t.loc("b")[1]  # New
```
#### Using index properties with a multi-column index
Previously an index on multiple columns was not usable in `.loc`, `.loc_indices` and `.iloc`. This PR fixes that.

#### General improvements
Each time I get into the index code, I find it confusing. So this PR continues efforts to expand documentation and make variable naming more consistent and expressive. Words like "key" are overloaded.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #13176

Closes https://github.com/astropy/astropy/pull/15826

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

mhvk commented on astropy/table/index.py:
```diff
@@ -975,22 +1011,58 @@ class TableLoc:
+        self.index_name = index_name
+
+    def __call__(self, *index_name):
+        if len(index_name) == 1 and isinstance(index_name[0], (tuple, list)):
```
Your example was `t.loc("b")[1]` - that would fail, correct? Was the plan just to pass on `index_name` (which is a tuple)?

mhvk reviewed (commented):
I think this is certainly much better, and the code is becoming clearer. My main comment is whether `index_name` is a good name, since it seems to have to be a `tuple` (unlike your example, so perhaps I just read too quickly...).

I also wondered about going for `__call__` altogether. The problem with `__getitem__` is that you cannot use keyword arguments, but with `__call__` this is no problem, i.e., one could do,
```
 t.loc(1, index="b")
```
The downside of that, though, is that one cannot easily pass in slices. Hence, your suggestion is probably better.

Another question would be whether one should be more explicit:
```
t.loc.using_index("b")[1]
```
Presumably it is fairly rare to have multiple indices in the first place?

neutrinoceros commented on astropy/table/index.py:
```diff
@@ -975,22 +1011,58 @@ class TableLoc:
     """
 
-    def __init__(self, table):
+    def __init__(self, table: Table, index_name: tuple | None = None):
```
supposedly ?
```suggestion
    def __init__(self, table: Table, index_name: str | tuple[str, ...] | None = None):
```

taldcroft replied:
Internally for this class, `index_id` must be `None` or `tuple[str, ...]`. I made the change.

neutrinoceros commented:
> This PR introduces a new syntax

`<pedantic review>` we're not defining a language, we have no power over *syntax*. What you're introducing here is a new public interface (API). `</pedantic review>`

> I also wondered about going for `__call__` altogether.

me too. Any class implementing `__call__` tends to behave weirdly and be abused, so I would prefer we go with an instance method with a clear name; Marten suggested `using_index`, which I can get behind, although I could also imagine a better name could be found.

neutrinoceros commented:
(other than that, I agree that the existing interface is quirky, confusing, and should be replaced with something better. Let's find out what it should be !)

taldcroft commented on astropy/table/index.py:
```diff
@@ -975,22 +1011,58 @@ class TableLoc:
+        Parameters
+        ----------
+        index_name : tuple
+            Name of the index to use
```
I also struggled with the name and was planning to revisit this. Fortunately `index_id` was high on my list and I think it is good since the name does not strongly imply a type so that `"a"` and `("a", "b")` both make sense.

Overall the strategy now is being liberal on input for an `index_id`, where it can be a string, tuple[str] or list[str]. This can simplify the calling code in places. We just need to be sure it doesn't introduce ambiguity or confusion.

taldcroft commented:
Well the precise meaning of syntax is beyond my pay grade , so I'll likely to continue abusing the word. :smile:

I went with `.with_index(..)`. Seems nice to me, almost like a context manager.

taldcroft commented:
I think I addressed the comments. Probably a little more cleanup, like `indexes` => `indices` everywhere. Gemini tells me that `indexes` is more common in the U.S. and Canada, and I personally like `indexes` better as not assuming anyone has studied latin, but the most important point now is consistency.

pllim commented:
Re: https://github.com/astropy/astropy/pull/18680#issuecomment-3371930291

SO also seems to prefer "indices" (https://stackoverflow.com/questions/1378781/proper-terminology-should-i-say-indexes-or-indices).

"Gemini" the AI, not the observatory, right? ;p

neutrinoceros commented:
for later reference, this supersedes #15826

taldcroft commented:
Thanks for the quick review again @mhvk and @neutrinoceros . 

I just pushed 3 more commits. I just made my first real protocol class for checking that the engine class is acceptable without any imports. Woohoo, but probably this needs checking...

The other two commits are just addressing the last of the review comments and a little one to improve the implementation of `add_index` a bit and to ensure that `tbl.primary_key` is always a tuple.

taldcroft commented on astropy/table/index.py:
```diff
@@ -188,13 +204,39 @@
+    """Protocol defining an index engine class"""
+
+    # Taken from soco.py
+    def __init__(self, data, row_index, unique=False): ...
```
I thought that protocols are a way to ensure that a class implements a behavior without doing so via class inheritance. E.g. a user writes their own index engine class and wants to use it here. As long as it follows the protocol then all good.

About extensive type hints, I see your point but here we might be in the "perfect is the enemy of good" territory. I have some more PR's in mind for this release and limited time, so I'd really like to get this merged without *too* much more effort. 

The use of a protocol here is confined to one little check, mostly to avoid a gotcha that I encountered where `t.add_index("a", "b")` did not raise any exceptions but was silently making a broken index. The alternative is explicitly importing all three astropy engine classes and checking for `issubclass` of one of those.

neutrinoceros replied:
> I'm not sure why Python would have included @runtime_checkable in the library if they did not intend for it to be used.

I'm not saying it should *never* be used, but rather, that it should be unnecessary in *new* code. Let me illustrate with a Venn diagram:
<img width="606" height="462" alt="Screenshot 2025-10-07 at 12 35 47" src="https://github.com/user-attachments/assets/686547d9-8a71-4cdc-b356-7e78b17c32c2" />

The main advantage of `Protocol`s is that they are static-analysis-time checkable, which in general is much better than *only* runtime-checkability as you don't need to run the code to know it's broken.
Now, if one's migrating existing code e.g., from runtime-registered abstract base classes, that are fundamentally runtime-checkable, to use protocols instead, making them runtime-checkable too is *needed* to avoid breaking backward compatibility. This is also the reason why I believe we should *only* use `runtime_checkable` when necessary: once a type is made runtime-checkable, we can't ever remove this property, even if it turns out difficult to maintain or otherwise problematic. If we instead start with pure protocols (no runtime checkability), we can always decide to add runtime-checkability later, but we shouldn't need it. That's one of the consequences of a fundamental principles for writing (more) maintanable code: only expose what's strictly necessary first, and add features as needed, never "just in case".


> The use of a protocol here is confined to one little check, mostly to avoid a gotcha that I encountered where t.add_index("a", "b") did not raise any exceptions but was silently making a broken index. The alternative is explicitly importing all three astropy engine classes and checking for issubclass of one of those.

I'm taking a bet there is at least one other approach to fixing this that doesn't involve runtime-checkable protocols. I'd be happy to take a deeper look and propose something less abstract later today.

> I have some more PR's in mind for this release and limited time, so I'd really like to get this merged without too much more effort.

I can make this review a priority on my end to make sure it gets the attention that it needs. And if that involves jumping in to add type-hints myself, that's certainly possible on my end, so please don't hesitate to ask for a hand whenever I'm asking too much; making changes myself isn't my default approach to review but I'm more than happy to help with the time I *do* have allocated to the project :)

neutrinoceros replied:
(oops, my Venn diagram is a bit inaccurate: explicit subclasses are, naturally, typecheck-time checkable)

neutrinoceros replied:
I confirm I'm fine with deferring this discussion to a later PR.

neutrinoceros commented on docs/changes/table/18680.api.rst:
```diff
@@ -0,0 +1,5 @@
+Selecting a table index in the ``.loc``, ``.iloc``, or ``.loc_indices`` properties by
+passing the index identifier as the first element of the item is deprecated and is
+planned for removal in astropy 9.0. For example, if a table ``t`` has two indices on
```
I don't think changelogs should contain this kind of promises.
```suggestion
passing the index identifier as the first element of the item is deprecated. For example, if a table ``t`` has two indices on
```

taldcroft replied:
I *completely* agree about not making promises about future plans (and this is something I notice all the time). 

However to me this is providing useful guidance to developers about the intended timeline. I was also careful to say "planned", which is not a promise. Note there is recent precedent from @mhvk which does make a stronger promise:
```
✗ more docs/changes/coordinates/18193.api.rst
The functionality of ``astropy.coordinates.concatenate`` and
``astropy.coordinates.concatenate_representations`` is now available using
``np.concatenate``. Hence, these functions are being deprecated, emitting an
``AstropyPendingDeprecationWarning`` starting with astropy 7.2. This will be
followed by a regular deprecation warning in astropy 8.0, and removal in 9.0.
```

taldcroft replied:
Another way to view this is not a promise but a warning... it's coming by 9.0, so you had better be ready.

neutrinoceros replied:
I missed that. I'll open another PR to propose dropping all such promises then.

mhvk replied:
I'm with @taldcroft here, in that for the changelog it is fine to give these warnings - they are read only at the time of release, almost certainly not by regular users, and give a sense that we're serious about it. I do agree, though, that in the documentation we should not add such statements. Rather, we should simply not use anything deprecated.

neutrinoceros commented on docs/changes/table/18680.feature.rst:
```diff
@@ -0,0 +1,10 @@
+which has two indices on columns ``"a"`` and ``"b"`` respectively,
+``t.loc.with_index("b")[2]`` will use index ``"b"`` to find all the table rows where
+``t["b"] == 2``. Doing this query using the previous syntax ``t.loc["b", 2]`` is
+deprecated and this functionality is planned for removal in astropy 9.0.
```
```suggestion
deprecated.
```

neutrinoceros commented on astropy/table/tests/test_index.py:
```diff
@@ -682,6 +682,17 @@ def test_index_id_item_deprecation_and_with_index(method, item):
+        TypeError,
+        match=r"engine must be an Engine class or instance, got 'b' instead.",
+    ):
+        t.add_index("a", "b")  # Easy mistake, too bad engine= is not keyword-only
```
it's a bit technical, but for reference, I think it's worth a shot (and I'm game if you's like me to give it a spin)

neutrinoceros commented on astropy/table/index.py:
```diff
@@ -967,6 +1027,20 @@ def __getitem__(self, item):
+""",
+)
+def interpret_item_as_index_id_and_item(item: tuple) -> tuple:
+    """Interpret the item as a (index_id, item) tuple."""
```
I find this confusing. Could you explain the difference between `item` (the input), and `item` (the last element on the output tuple) ?
I can include the explanation later if you don't want to spin CI again here :)

taldcroft replied:
This is exactly the confusion that is being deprecated away. You can do `t.loc[index_id, actual_item]` which translates to `t.loc.with_index(index_id)[actual_item`. But before that we have `t.loc.__getitem__(item)`, and that is the input `item` here.

The only reason for this little function is so that it can be deprecated now and then removed in 9.0. Previously this was just a line of code in an `if` block.
